Assets/LeapMotion/Scripts/DataStructures/Editor/Tests/ProduceConsumeBufferTest.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Threading;
+using NUnit.Framework;
+
+namespace Leap.Unity.Tests {
+
+  public class ProduceConsumeBufferTest {
+
+    private ProduceConsumeBuffer<TestStruct> buffer;
+
+    [SetUp]
+    public void Setup() {
+      buffer = new ProduceConsumeBuffer<TestStruct>(16);
+    }
+
+    [TearDown]
+    public void Teardown() {
+      buffer = null;
+    }
+
+    [Test]
+    [Timeout(1000)]
+    public void Test() {
+      Thread consumer = new Thread(new ThreadStart(consumerThread));
+      Thread producer = new Thread(new ThreadStart(producerThread));
+
+      consumer.Start();
+      producer.Start();
+
+      consumer.Join();
+      producer.Join();
+    }
+
+    private void consumerThread() {
```

Comment:
Since the increment of the head or tail index is done *after* the write has been finished, and since the assignment of the new value is atomic, the data structure cannot ever be left in an invalid state.